File: 337-house-robber-iii/house-robber-iii.py
class Solution:
    @cache
    def rob(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        x=root.val
        if root.left != None :
            x+=self.rob(root.left.left)+self.rob(root.left.right)
        if root.right != None :
            x+=self.rob(root.right.left)+self.rob(root.right.right)
        return max(x,self.rob(root.left)+self.rob(root.right))

File: 337-house-robber-iii/house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    unordered_map<TreeNode*,int> dp;
public:
    int rob(TreeNode* root) {
        if(!root) return 0;
        if(dp.count(root)) return dp[root];
        return dp[root]=max(
            rob(root->left)+rob(root->right),
            root->val
            +(root->left?rob(root->left->right)+rob(root->left->left):0)
            +(root->right?rob(root->right->right)+rob(root->right->left):0));
    }
};